model/pipeline/src/main/proto/beam_runner_api.proto

```diff
@@ -489,33 +490,66 @@ message TestStreamPayload {
   // (Required) the coder for elements in the TestStream events
   string coder_id = 1;
 
+  // (Optional) If specified, the TestStream will replay these events.
   repeated Event events = 2;
 
+  // (Optional) If specified, the TestStream will use the InteractiveService
+  // protocol found in beam_interactive_api.proto to retrieve events.
+  ApiServiceDescriptor endpoint = 3;
+
   message Event {
     oneof event {
       AdvanceWatermark watermark_event = 1;
       AdvanceProcessingTime processing_time_event = 2;
       AddElements element_event = 3;
     }
 
+    // Advances the watermark to the specified timestamp.
     message AdvanceWatermark {
+      // (Required) The watermark to advance to.
       int64 new_watermark = 1;
+
+      // (Optional) The output watermark tag for a PCollection. If unspecified
+      // or with an empty string, this will default to the Main PCollection
+      // Output
+      string tag = 2;
     }
 
+    // Advances the processing time clock by the specified amount.
     message AdvanceProcessingTime {
+      // (Required) The duration to advance by.
       int64 advance_duration = 1;
     }
 
+    // Adds elements to the stream to be emitted.
     message AddElements {
+      // (Required) The elements to add to the TestStream.
       repeated TimestampedElement elements = 1;
+
+      // (Optional) The output PCollection tag to add these elements to. If
+      // unspecified or with an empty string, this will default to the Main
+      // PCollection Output.
+      string tag = 3;
     }
   }
 
+  // A single element inside of the TestStream.
   message TimestampedElement {
+    // (Required) The element encoded. Currently the TestStream only supports
+    // encoding primitives.
     bytes encoded_element = 1;
+
+    // (Required) The event timestamp of this element.
     int64 timestamp = 2;
   }
 }
+
+service TestStreamService {
```

Comment:
Why is this needed?

The current ProcessBundleDescriptor and ProcessBundleRequest can already split the execution so that the Runner and the SDK perform their respective responsibilities:
```
Runner that is processing and emitting TestStream events -> Data Channel -> SDK processing events
```